core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -124,15 +134,73 @@ object DumpLogSegments {
         println("offset: %d position: %d".format(entry.offset + index.baseOffset, entry.position))
     }
   }
-  
+
+  private trait MessageParser {
+    def parse(message: Message): (Option[_], Option[_])
```

Comment:
Agreed that it's a little weird. I had the type parameters initially, but I forgot why I had to abandon them. For what it's worth, we don't actually get much value from them since (as you noticed) we're just calling toString on the result (and there's not much else we can do within the constraints of Decoder). Let me try to add them back because it's still probably cleaner.